File: dotnet/Multiplayer/Rooms/Actions/Actions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Michitai.Multiplayer.Rooms.Actions
{
    /// <summary>
    /// Provides static methods for managing room actions.
    /// Handles submitting actions, polling for completed actions, retrieving pending actions, and completing actions.
    /// </summary>
    public class Actions
    {
        /// <summary>
        /// Submits an action to target players in the room.
        /// </summary>
        /// <typeparam name="T">The type of action data.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="request">The action submission request containing targets and data.</param>
        /// <param name="ct">Cancellation token.</param>
        /// <returns>Task containing the action submission response.</returns>
        public static Task<ActionSubmitResponse> SubmitActionAsync<T>(Client client, string playerToken, SubmitAction<T> request, CancellationToken ct = default) where T : class, new()
            => client.Send<ActionSubmitResponse>(HttpMethod.Post, client.Url(Endpoints.GameRoomActions,
        $"&player_token={playerToken}"), new ActionSubmitRequest<T>(request.Target_players, request.Action_type, request.Request_data, request.Target_players_ids), ct);

        /// <summary>
        /// Polls for completed actions that were targeted to the current player.
        /// </summary>
        /// <typeparam name="T">The type of action data.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="ct">Cancellation token.</param>
        /// <returns>Task containing the poll actions response with completed actions.</returns>
        public static Task<ActionPollResponse<T>> PollActionsAsync<T>(Client client, string playerToken, CancellationToken ct = default) where T : class, new()
            => client.Send<ActionPollResponse<T>>(HttpMethod.Get, client.Url(Endpoints.GameRoomActionsPoll, $"&player_token={playerToken}"), null, ct);

        /// <summary>
        /// Retrieves pending actions that need to be completed by the host.
        /// Only the host can retrieve pending actions.
        /// </summary>
        /// <typeparam name="T">The type of action data.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="ct">Cancellation token.</param>
        /// <returns>Task containing the pending actions response.</returns>
        public static Task<ActionPendingResponse<T>> GetPendingActionsAsync<T>(Client client, string playerToken, CancellationToken ct = default) where T : class, new()
            => client.Send<ActionPendingResponse<T>>(HttpMethod.Get, client.Url(Endpoints.GameRoomActionsPending, $"&player_token={playerToken}"), null, ct);
        
        /// <summary>
        /// Marks an action as complete with an optional response.
        /// Only the host can complete actions.
        /// </summary>
        /// <typeparam name="T">The type of response data.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="actionId">The ID of the action to complete.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="request">The action completion request containing status and response data.</param>
        /// <param name="ct">Cancellation token.</param>
        /// <returns>Task containing the action completion response.</returns>
        public static Task<ActionCompleteResponse> CompleteActionAsync<T>(Client client, string actionId, string playerToken,
            ActionComplete<T> request, CancellationToken ct = default) where T : class, new()
            => client.Send<ActionCompleteResponse>(HttpMethod.Post, client.Url(string.Format(Endpoints.GameRoomActionComplete, actionId),
                $"&player_token={playerToken}"), new ActionCompleteRequest<T>(request.Status, request.Response_data), ct);
    }
}

File: python/multiplayer/rooms/actions/actions.py
"""Room actions: submit, poll completed, list pending, complete (host)."""

from __future__ import annotations

import dataclasses
from typing import Any, Optional, Type, TypeVar

from ..._serialization import from_dict
from ...client import Client
from ..enums import RoomTargetPlayers
from .enums import RoomCompleteActionStatus
from .models import (
    ActionCompleteResponse,
    ActionPendingResponse,
    ActionPollResponse,
    ActionSubmitResponse,
)

T = TypeVar("T")

# Endpoints
GAME_ROOM_ACTIONS = "game_room.php/actions"
GAME_ROOM_ACTIONS_POLL = "game_room.php/actions/poll"
GAME_ROOM_ACTIONS_PENDING = "game_room.php/actions/pending"
GAME_ROOM_ACTION_COMPLETE = "game_room.php/actions/{action_id}/complete"


class Actions:
    """Manage room actions between players and the host."""

    def __init__(self, client: Client) -> None:
        self._client = client

    def submit(
        self,
        player_token: str,
        action_type: str,
        request_data: Any = None,
        target_players: RoomTargetPlayers = RoomTargetPlayers.ALL,
        target_players_ids: Optional[list] = None,
    ) -> ActionSubmitResponse:
        """Submit an action to target players in the room."""
        return self._client.send(
            "POST",
            self._client.player_url(GAME_ROOM_ACTIONS, player_token),
            {
                "target_players": target_players.value,
                "target_players_ids": target_players_ids,
                "action_type": action_type,
                "request_data": request_data,
            },
            ActionSubmitResponse,
        )

    def poll(
        self,
        player_token: str,
        data_cls: Optional[Type[T]] = None,
    ) -> ActionPollResponse:
        """Poll completed actions that were targeted to the current player.

        :param data_cls: Optional dataclass type for each action's ``response_data``.
        """
        response = self._client.send(
            "GET",
            self._client.player_url(GAME_ROOM_ACTIONS_POLL, player_token),
            None,
            ActionPollResponse,
        )
        _apply_type(response.actions, "response_data", data_cls)
        return response

    def get_pending(
        self,
        player_token: str,
        data_cls: Optional[Type[T]] = None,
    ) -> ActionPendingResponse:
        """Get pending actions to complete (host only).

        :param data_cls: Optional dataclass type for each action's ``request_data``.
        """
        response = self._client.send(
            "GET",
            self._client.player_url(GAME_ROOM_ACTIONS_PENDING, player_token),
            None,
            ActionPendingResponse,
        )
        _apply_type(response.actions, "request_data", data_cls)
        return response

    def complete(
        self,
        player_token: str,
        action_id: str,
        status: RoomCompleteActionStatus = RoomCompleteActionStatus.COMPLETED,
        response_data: Any = None,
    ) -> ActionCompleteResponse:
        """Mark an action as complete with an optional response (host only)."""
        return self._client.send(
            "POST",
            self._client.player_url(
                GAME_ROOM_ACTION_COMPLETE.format(action_id=action_id), player_token
            ),
            {"status": status.value, "response_data": response_data},
            ActionCompleteResponse,
        )


def _apply_type(items: Any, attr: str, cls: Optional[Type[T]]) -> None:
    if cls is None:
        return
    for item in items:
        value = getattr(item, attr)
        if dataclasses.is_dataclass(cls) and isinstance(value, dict):
            setattr(item, attr, from_dict(cls, value))
